File: app/src/main/java/ru/org/adons/cblock/ui/viewmodel/MainViewModel.java
package ru.org.adons.cblock.ui.viewmodel;

import android.content.Context;

import java.util.List;

import javax.inject.Inject;

import ru.org.adons.cblock.R;
import ru.org.adons.cblock.app.Preferences;
import ru.org.adons.cblock.data.BlockListModel;
import ru.org.adons.cblock.model.BlockListItem;
import ru.org.adons.cblock.service.BlockService;
import ru.org.adons.cblock.ui.view.MainFragment;
import ru.org.adons.cblock.utils.Logging;
import rx.Observable;

/**
 * Main Model View, provide data for {@link MainFragment}
 */
public class MainViewModel {

    @Inject Context context;
    @Inject Preferences pref;
    @Inject BlockListModel blockListModel;

    public Observable<List<BlockListItem>> getBlockList() {
        return blockListModel.getBlockList()
                .doOnSubscribe(Logging.subscribe(this.getClass(), "getBlockList"))
                .doOnUnsubscribe(Logging.unsubscribe(this.getClass(), "getBlockList"));
    }

    public Observable<Boolean> getSwitchState() {
        return Observable.just(Preferences.SERVICE_SWITCH_KEY).map(pref::getBoolean);
    }

    public Observable<String> changeServiceState(boolean isEnable) {
        return Observable.just(isEnable).map(bool -> {
            pref.putBoolean(Preferences.SERVICE_SWITCH_KEY, isEnable);
            if (bool){
                BlockService.enable(context);
                return context.getString(R.string.main_notification_text_enable);
            } else {
                BlockService.disable(context);
                return context.getString(R.string.main_toast_text_disable);
            }
        });
    }

}

File: app/src/main/java/mobile/addons/cblock/ui/view/main/MainViewModel.kt
package mobile.addons.cblock.ui.view.main

import android.content.Context
import mobile.addons.cblock.R
import mobile.addons.cblock.app.BlockManager
import mobile.addons.cblock.app.Preferences
import mobile.addons.cblock.data.BlockListItem
import mobile.addons.cblock.ext.logSubscribe
import mobile.addons.cblock.ext.logUnsubscribe
import mobile.addons.cblock.model.BlockListModel
import mobile.addons.cblock.service.BlockService
import rx.Observable
import javax.inject.Inject

/**
 * Main Model View, provide data for [MainFragment]
 */

class MainViewModel @Inject constructor() {

    @Inject lateinit var context: Context
    @Inject lateinit var pref: Preferences
    @Inject lateinit var blockListModel: BlockListModel
    @Inject lateinit var blockManager: BlockManager

    fun getBlockList(): Observable<List<BlockListItem>> = blockListModel.getBlockList()
            .doOnSubscribe { logSubscribe("getBlockList") }
            .doOnUnsubscribe { logUnsubscribe("getBlockList") }

    fun getServiceState(): Observable<Boolean> = Observable.fromCallable { pref.serviceState }

    fun changeServiceState(isEnable: Boolean): Observable<String> = Observable
            .just(isEnable)
            .map {
                pref.serviceState = it
                if (it) {
                    BlockService.start(context)
                    context.getString(R.string.main_notification_text_enable)
                } else {
                    BlockService.stop(context)
                    context.getString(R.string.main_toast_text_disable)
                }
            }

    fun deletePhone(itemId: Long) = blockManager.deletePhone(blockListModel, itemId)

}
